File: Year-2023/LeetCode-July-Challenge/LRUCache.cpp
#include<bits/stdc++.h>
using namespace std;
class LRUCache {
private:
    int capacity;
    queue<int> history;
    unordered_map<int,int> counter;
    unordered_map<int,int> dict;
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }

    int get(int key) {
        if(dict.count(key)){
            history.push(key);
            counter[key]++;
            return dict[key];
        }
        return -1;
    }

    void removeHistory(){
        while(!history.empty()){
            int curr = history.front();
            history.pop();
            counter[curr]--;
            if(counter[curr] == 0){
                dict.erase(curr);
                break;
            }
        }
        return;
    }
    void put(int key, int value) {
        history.push(key);
        counter[key]++;
        dict[key] = value;
        if(dict.size() > this->capacity) removeHistory();
        return;
    }
};

int main(){
    LRUCache *l = new LRUCache(2);
    l->put(1,1);
    l->put(2,2);
    cout<<l->get(1)<<endl;
    return 0;
}

File: Year-2023/LeetCode-July-Challenge/LRUCache.java
import java.util.*;

public class LRUCache {
    private int capacity;
    private Queue<Integer> history = new LinkedList<>();
    private HashMap<Integer, Integer> counter = new HashMap<>();
    private HashMap<Integer, Integer> dict = new HashMap<>();

    public LRUCache(int capacity) {
        this.capacity = capacity;
    }

    public int get(int key) {
        if(dict.containsKey(key)){
            history.add(key);
            counter.put(key, counter.getOrDefault(key, 0) + 1);
            return dict.get(key);
        }
        return -1;
    }

    private void removeHistory(){
        while(!history.isEmpty()){
            int curr = history.poll();
            counter.put(curr, counter.get(curr) - 1);
            if(counter.get(curr) == 0){
                dict.remove(curr);
                break;
            }
        }
    }

    public void put(int key, int value) {
        history.add(key);
        counter.put(key, counter.getOrDefault(key, 0) + 1);
        dict.put(key, value);
        if(dict.size() > this.capacity) removeHistory();
        return;
    }
    public static void main(String[] args) {
        LRUCache l = new LRUCache(2);
        l.put(1,1);
        l.put(2,2);
        System.out.println(l.get(1));
    }
}
